File: docs/article/source/scala/apps/cargo-tracker/app/cargotracker/auth/domain/model/valueobjects/Role.scala
package cargotracker.auth.domain.model.valueobjects

/** システムユーザーの役割（ロール）。
  *
  * 各ロールは UI ナビゲーション・認可・業務ログの分離に利用する。
  */
enum Role:
  case Sales
  case RouteDesigner
  case Tracker
  case Settlement
  case MasterAdmin

object Role:

  /** 永続化からの読み戻し用に文字列名と Role を相互変換する。
    *
    * 未知の名前は [[scala.None]] を返す。
    */
  def fromName(name: String): Option[Role] =
    values.find(_.toString.equalsIgnoreCase(name))

File: apps/cargo-tracker/src/main/java/com/example/cargotracker/security/domain/model/valueobjects/Role.java
package com.example.cargotracker.security.domain.model.valueobjects;

/**
 * システムのロール。
 *
 * <p>定義の正典は {@code docs/design/non_functional.md} の RBAC ロール定義である。
 * 値を追加・変更する場合は先にそちらを更新すること。
 */
public enum Role {
    /** 営業担当者。見積・荷主・予約を扱う。 */
    SALES,
    /** 荷主。自分の予約と追跡を参照する。 */
    SHIPPER,
    /** 荷受人。追跡を参照する。 */
    CONSIGNEE,
    /** 経路設計者。航路と経路割り当てを扱う。 */
    ROUTER,
    /** 追跡管理者。追跡・例外・通関を扱う。 */
    TRACKER,
    /** 荷役作業員。荷役と通関を扱う。 */
    HANDLER,
    /** 経理担当者。請求と精算を扱う。 */
    BILLING,
    /** 管理者。 */
    ADMIN;

    /** Spring Security が用いる {@code ROLE_} 接頭辞付きの権限名を返す。 */
    public String authority() {
        return "ROLE_" + name();
    }
}
